cnd/src/network/start_swap.rs

```diff
@@ -0,0 +1,90 @@
+use crate::{
+    db::{ForSwap, Save, Swap},
+    network::comit::RemoteData,
+    protocol_spawner::{ProtocolSpawner, Spawn},
+    storage::{Load, Storage},
+    swap_protocols::{halight, herc20, LocalSwapId},
+};
+use ::comit::{
+    network::{WhatAliceLearnedFromBob, WhatBobLearnedFromAlice},
+    Protocol, Role, Side,
+};
+use chrono::offset::Utc;
+
+#[allow(clippy::too_many_arguments)]
+pub async fn start_swap(
+    storage: Storage,
+    spawner: ProtocolSpawner,
+    id: LocalSwapId,
+    data: RemoteData,
+) -> anyhow::Result<()>
+where
+    ProtocolSpawner: Spawn<herc20::Params> + Spawn<halight::Params>,
+{
+    let start_of_swap = Utc::now().naive_local();
+    let swap = storage.load(id).await?;
+
+    match (swap, data) {
+        (
+            Swap {
+                alpha: Protocol::Herc20,
+                beta: Protocol::Halight,
+                role: role @ Role::Alice,
+            },
+            RemoteData {
+                ethereum_identity: Some(ethereum_identity),
+                lightning_identity: Some(lightning_identity),
+                secret_hash: None,
+            },
+        ) => {
+            spawner.supports_halight()?;
+            storage
+                .save(ForSwap {
+                    local_swap_id: id,
+                    data: WhatAliceLearnedFromBob {
+                        redeem_ethereum_identity: ethereum_identity,
+                        refund_lightning_identity: lightning_identity,
+                    },
+                })
+                .await?;
+
+            let herc20_params: herc20::Params = storage.load(id).await?;
+            let halight_params: halight::Params = storage.load(id).await?;
```

Comment:
Over in `respawn`, I am loading these in one call. Not sure what the better approach is.

As we can see here, doing them separately might allow us to create an API where you just pass `id`, load everything else from the DB and spawn. That could be a useful abstraction on top of `ProtocolSpawner`. Let's keep the idea in mind :)

Reply:
Copied the code as suggested.